mlir-sycl/doc/design/SYCLGridRepresentation.md

```diff
@@ -0,0 +1,244 @@
+
+```MLIR
+%0 = sycl.call() {FunctionName = @get_global_id, MangledFunctionName = @..., TypeName = @nd_item} : (!memref<?xsycl_nd_item_1_>) -> !sycl_id_1_
+%1 = sycl.call() {FunctionName = @get_local_id, MangledFunctionName = @..., TypeName = @nd_item} : (!memref<?xsycl_nd_item_1_>) -> !sycl_id_1_
+```
+
+We can see how, even though `%0` and `%1` have a very different nature, they
+have the same type and defining operation, in fact, they only differ on the
+`FunctionName` (and `MangledFunctionName`) attribute of their defining
+operations. As the target is to make memory access analysis easier, we propose
+index space identifiers of different nature should differ at least on their
+defining operation.
+
+Also, in order to perform memory access analysis, these operations should be
+presented in a canonical form. A canonicalization pattern checking every
+`sycl.call` operation in the MLIR module and replacing it with different
+`sycl.call` operations can be a bit cumbersome, thus we propose these operations
+to have different defining operations.
+
+## `SYCLMethodOpInterface`
+
+As these operations are really similar in terms of structure, the
+`SYCLMethodOpInterface` operation interface is defined as the interface
+describing an operation representing a member function call of a SYCL type. This
+interface provides the following functions:
+
+| Function                                                       | Description                                                            |
+|:---------------------------------------------------------------|:-----------------------------------------------------------------------|
+| `static mlir::TypeID getTypeID();`                             | Return the ID of the type this method is a member of.                  |
+| `static llvm::ArrayRef<llvm::StringLiteral> getMethodNames();` | Return the list of the method names to be replaced by this operation.  |
+| `llvm::ArrayRef<mlir::Type> getFunctionArgTypes();`            | Return the argument types of the function implementing this operation. |
+| `llvm::StringRef getFunctionName();`                           | Return the name of the function implementing this operation.           |
+| `llvm::StringRef getTypeName();`                               | Return the original name of the type this method is implemented in.    |
+
+As we can see, the attributes `FunctionName` and `TypeName` are analogous to the
+attributes present in the `sycl.call` operation, as this interface also
+represents function calls. On the other hand, the `FunctionArgTypes` attribute
+contains the argument types of the member function to call. E.g., in an
+implementation in which `nd_item` was derived from a base class and the generic
+SPIR-V address space was to be used for pointer arguments (an implementation
+different from the one corresponding to the example above), a function call to
+`nd_item::get_global_id()` would have `{!llvm.ptr<struct<...>, 4>}` as
```

Comment:
Note that we don't use a SYCL type here as the base type may not be in our dialect.